protocols/raft/src/main/java/io/atomix/protocols/raft/impl/RaftContext.java

```diff
@@ -538,11 +538,18 @@ public void transition(RaftServer.Role role) {
     try {
       this.role = createRole(role);
       this.role.start().get();
+
     } catch (InterruptedException | ExecutionException e) {
       throw new IllegalStateException("failed to initialize Raft state", e);
     }
 
-    roleChangeListeners.forEach(l -> l.accept(this.role.role()));
+    if (this.role.role() == role) {
```

Comment:
Did you run into this? :fearful:

Reply:
Yes. It can happen on single nodes because they directly translate to Leader on follower :see_no_evil: